File: molecules/calculator.py
from typing import List
from particle import Particle
from vector import Vector2D
from time import time
from particles_factory import ParticlesCells
import math
import numpy as np
class Calculator:
    def __init__(self, width, height, initial_particles, **kwargs):
        self.width = width
        self.height = height
        self.delta_time = kwargs.get('delta_time', 0.001)
        self.cut_off_distance = kwargs.get('cut_off_distance', 2.5)

    def limit_condition_periodic(self, particle):
        particle.center.x %= self.width
        particle.center.y %= self.height

    def iteration(self, particles: ParticlesCells, log_time: bool):
        t1 = time()
        for particle in particles.iterate_throw_particles():
            particle.velocity += particle.acceleration * (self.delta_time / 2)
            particle.translate(particle.velocity * self.delta_time)
            self.limit_condition_periodic(particle)
        particles.update_cells()
        t_before = time()
        self.drop_acceleration_and_potential(particles)
        for particle1, particle2 in particles.iterate_throw_particle_pairs():
            self.recompute_acceleration_and_potential(particle1, particle2)
        t_after = time()

        for particle in particles.iterate_throw_particles():
            particle.velocity += particle.acceleration * (self.delta_time / 2)
        t2 = time()
        if log_time:
            print(f'Total={t2 - t1}. Acceleration = {t_after - t_before}, F1 = {t_before - t1}, F2 = {t2 - t_after}')

    def iteration_2(self, particles: ParticlesCells, log_time: bool):
        t1 = time()
        for particle in particles.iterate_throw_particles():
            particle.translate(particle.velocity * self.delta_time + particle.acceleration * (self.delta_time ** 2 / 2))
            particle.velocity += particle.acceleration * (self.delta_time / 2)
            self.limit_condition_periodic(particle)
        t11 = time()
        particles.update_cells()
        t_before = time()
        self.drop_acceleration_and_potential(particles)
        for particle1, particle2 in particles.iterate_throw_particle_pairs():
            self.recompute_acceleration_and_potential(particle1, particle2)
        t_after = time()

        for particle in particles.iterate_throw_particles():
            particle.velocity += particle.acceleration * (self.delta_time / 2)
        t2 = time()
        if log_time:
            print(f'Total={t2 - t1}. Acceleration = {t_after - t_before}, F1 = {t11 - t1}, F2 = {t_before - t11}')

    def recompute_acceleration_and_potential(self, particle_i: Particle, particle_j: Particle):
        radius_i_j = particle_j.center - particle_i.center

        if abs(radius_i_j.x) > (self.width / 2):
            radius_i_j.x = -(self.width - abs(radius_i_j.x)) * np.sign(radius_i_j.x)
        if abs(radius_i_j.y) > (self.height / 2):
            radius_i_j.y = -(self.height - abs(radius_i_j.y)) * np.sign(radius_i_j.y)
        radius_i_j_scalar = abs(radius_i_j)

        if radius_i_j_scalar < self.cut_off_distance:
            radius_i_j_inverted = 1 / radius_i_j_scalar
            acceleration_i_j = radius_i_j * (-48 * ((radius_i_j_inverted ** 14) - 0.5 * (radius_i_j_inverted ** 8)))
            potential_i_j = 4 * (radius_i_j_inverted**12 - radius_i_j_inverted**6)

            particle_i.acceleration += acceleration_i_j
            particle_j.acceleration -= acceleration_i_j

            particle_i.potential += potential_i_j
            particle_j.potential += potential_i_j

    def drop_acceleration_and_potential(self, particles: ParticlesCells):
        for p in particles.iterate_throw_particles():
            p.acceleration = Vector2D()
            p.potential = 0.0


def get_width_height(particles_count, density):
    length = (particles_count / density) ** 0.5
    return (length, length)


def calculate_center_of_mass_velocity(particles: ParticlesCells):
    # P = M*Vc
    total_momentum = sum([np.multiply(p.velocity.coordinates, p.mass) for p in particles.iterate_throw_particles()])
    total_mass = sum([p.mass for p in particles.iterate_throw_particles()])

    return total_momentum * (1 / total_mass)

File: molecules_cpp/Calculator.cpp
//
// Created by marka on 25.10.2019.
//

#include "Calculator.h"

Calculator::Calculator(int number_of_particles, long double density, long double delta_time) {
    this->number_of_particles = number_of_particles;
    this->delta_time = delta_time;
    this->box_width = Calculator::calculate_box_width(number_of_particles, density);
}

int Calculator::signum(long double x) {
    if (x == 0.0) {
        return 0;
    } else if (x < 0) {
        return -1;
    } else {
        return 1;
    }
}

void Calculator::recompute_acceleration_and_potential(Particle &p1, Particle &p2) {
    // std::cout << p1.id << " " << p2.id << std::endl;

    Vector2D radius_vector = p2.center - p1.center;
    if (abs(radius_vector.x) > this->box_width / 2) {
        radius_vector.x = -(this->box_width - abs(radius_vector.x)) * Calculator::signum(radius_vector.x);
    }
    if (abs(radius_vector.y) > this->box_width / 2) {
        radius_vector.y = -(this->box_width- abs(radius_vector.y)) * Calculator::signum(radius_vector.y);
    }
    long double radius_vector_scalar = radius_vector.get_norm();

    long double radius_vector_inverted = 1 / radius_vector_scalar;
    long double rr = radius_vector_inverted * radius_vector_inverted;
    long double rrr = radius_vector_inverted * rr;
    long double r6 = rrr * rrr;
    long double r12 = r6 * r6;
    Vector2D acceleration = radius_vector * (-48 * (r12 * rr - 0.5 * r6 * rr));
    long double potential = 2 * (r12 - r6);
    p1.acceleration = p1.acceleration + acceleration;
    p2.acceleration = p2.acceleration - acceleration;

    p1.potential += potential;
    p2.potential += potential;
}

void Calculator::limit_condition_periodic(Particle &particle) {
    if (particle.center.x >= box_width) {
        particle.center.x -= box_width;
    } else if (particle.center.x < 0) {
        particle.center.x += box_width;
    }
    if (particle.center.y >= box_width) {
        particle.center.y -= box_width;
    } else if (particle.center.y < 0) {
        particle.center.y += box_width;
    }
}

void Calculator::iteration(std::vector<Particle> &particles) {
    for (auto & particle : particles) {
        particle.center = particle.center + particle.velocity * this->delta_time + particle.acceleration * (this->delta_time * this->delta_time / 2);
        particle.velocity = particle.velocity + particle.acceleration * (this->delta_time / 2);
        limit_condition_periodic(particle);
        Calculator::drop_potential_and_acceleration(particle);
    }

    for (int i = 0; i < this->number_of_particles - 1; i++) {
        for (int j = i+1; j < this->number_of_particles; j++) {
            if (particles[i].id != particles[j].id)
                recompute_acceleration_and_potential(particles[i], particles[j]);
        }
    }
    for (auto & particle : particles) {
        particle.velocity = particle.velocity + particle.acceleration * (this->delta_time / 2);
    }
}

void Calculator::iteration(ParticlesCells &cells) {
    for (int i = 0; i < cells.cells_side_count; i++) {
        for (int j = 0; j < cells.cells_side_count; j++) {
            auto & cell_particles = cells.cells[i][j];
            for (auto & particle : cell_particles) {
                particle.center = particle.center + particle.velocity * this->delta_time + particle.acceleration * (this->delta_time * this->delta_time / 2);
                particle.velocity = particle.velocity + particle.acceleration * (this->delta_time / 2);
                limit_condition_periodic(particle);
                Calculator::drop_potential_and_acceleration(particle);
            }
        }
    }

    cells.update_cells();

    for (int i = 0; i < cells.cells_side_count; i++) {
        for (int j = 0; j < cells.cells_side_count; j++) {
            auto & cell1_particles = cells.cells[i][j];

            for (int particle1_index = 0; particle1_index + 1 < cell1_particles.size(); particle1_index++) {
                for (int particle2_index = particle1_index + 1; particle2_index < cell1_particles.size(); particle2_index++) {
                    auto & p1 = cell1_particles[particle1_index];
                    auto & p2 = cell1_particles[particle2_index];

                    recompute_acceleration_and_potential(p1, p2);
                }
            }

            for (auto & particle1 : cell1_particles) {
                for (int cell_di = -1; cell_di <= 1; cell_di++) {
                    for (int cell_dj = -1; cell_dj <= 1; cell_dj++) {
                        if (cell_di == 0 && cell_dj == 0) {
                            continue;
                        }

                        int adjacent_cell_i = i + cell_di;
                        int adjacent_cell_j = j + cell_dj;

                        if (adjacent_cell_i < 0) adjacent_cell_i += cells.cells_side_count;
                        if (adjacent_cell_i >= cells.cells_side_count) adjacent_cell_i -= cells.cells_side_count;
                        if (adjacent_cell_j < 0) adjacent_cell_j += cells.cells_side_count;
                        if (adjacent_cell_j >= cells.cells_side_count) adjacent_cell_j -= cells.cells_side_count;

                        if (adjacent_cell_i < i || (adjacent_cell_i == i && adjacent_cell_j < j)) {
                            continue;
                        }

                        auto & cell2_particles = cells.cells[adjacent_cell_i][adjacent_cell_j];
                        for (auto & particle2 : cell2_particles) {
                            recompute_acceleration_and_potential(particle1, particle2);
                        }
                    }
                }
            }
        }
    }

    for (int i = 0; i < cells.cells_side_count; i++) {
        for (int j = 0; j < cells.cells_side_count; j++) {
            auto & cell_particles = cells.cells[i][j];
            for (auto & particle : cell_particles) {
                particle.velocity = particle.velocity + particle.acceleration * (this->delta_time / 2);
            }
        }
    }
}

void Calculator::drop_potential_and_acceleration(Particle & particle) {
    particle.potential = 0.0;
    particle.acceleration = Vector2D(0.0, 0.0);
}

long double Calculator::calculate_box_width(int number_of_particles, long double density) {
    long double length = sqrtf(number_of_particles) / sqrtf(density);
    return length;
}

